### PR TITLE
simplest change to prevent exposing SinkRef => SourceRef => SinkRef cycle

### PR DESCRIPTION
**Pull request by jrudolph**
That's the smallest set of changes I could come up with for preventing internal details leaking into the API. I think the separation between API and implementation should even be extended further.

### CODE REVIEW

**Comment by jrudolph** [comment]
I added another commit to solve the javadsl/scaladsl problem:

 * Provide the API (`SourceRef`/`SinkRef`) traits directly in `akka.stream`
 * They have `getSource` / `getSink` methods to access the java version (is that naming fine or should it be more explicit, like `javaSource` / `javaSink`.
 * I created another commit moving the constructors to `{javadsl,scaladsl}.{Source,Sink}`, so it's now `Source.sinkRef`, instead of `SinkRef.source` (same for java and scala)

**Comment by jrudolph** [comment]
Added two more commits:
 * removed canMaterializeXYZ which is not needed any more
 * split up stage (implementation) and ref (data to be serialized) so it doesn't look as if the stage itself is serialized

**Comment by jrudolph** [comment]
Added another change, that made the ActorRef non-optional in the Source/SinkRefImpl to clarify that the ref is never optional in the messages sent around on the wire.

All these changes are pure refactorings so far. Please take what you like, no need to keep any of the commits.

File: akka-stream/src/main/protobuf/StreamRefContainers.proto [hunk 1]
```diff
@@ -12,17 +12,11 @@ option optimize_for = SPEED;
 // TODO make explicit types for SinkRef/SourceRef as they can carry initial settings (initial demand etc)
 
 message SinkRef {
-    optional ActorRef targetRef = 1;
-    // this will be basically always false, since we already materialized this "connection"
-    // so materializing another Source, where this Sink was created BY materializing a Source does not make sense
-    optional bool canMaterializeSource = 2;
+    ActorRef targetRef = 1;
```
Comment (jrudolph) [inline akka-stream/src/main/protobuf/StreamRefContainers.proto:15]: Protobuf files need to be regenerated, I didn't try to make that work on my machine.

Comment (ktoso) [reply akka-stream/src/main/protobuf/StreamRefContainers.proto:15]: Whoop, will do

File: akka-stream-tests/src/test/scala/akka/stream/remote/StreamRefsSpec.scala [hunk 2]
```diff
@@ -42,31 +42,31 @@ object StreamRefsSpec {
          * For them it's a Source; for us it is a Sink we run data "into"
          */
         val source: Source[String, NotUsed] = Source(List("hello", "world"))
-        val ref: Future[SourceRef[String]] = source.runWith(SourceRef.sink())
+        val ref: Future[SourceRef[String]] = source.runWith(Sink.sourceRef())
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/remote/StreamRefsSpec.scala:45]: Awesome idea! Reads much better and discoverable <3

File: akka-stream/src/main/protobuf/StreamRefContainers.proto (hunk 1 above)
Comment (ktoso) [reply akka-stream/src/main/protobuf/StreamRefContainers.proto:15]: I think the required keyword is needed tho? Yes, excellent that we can have it required :)

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 3]
```diff
@@ -278,6 +279,9 @@ object Sink {
     new Sink(scaladsl.Sink.lazyInit[T, M](
       t ⇒ sinkFactory.apply(t).toScala.map(_.asScala)(ExecutionContexts.sameThreadExecutionContext),
       () ⇒ fallback.create()).mapMaterializedValue(_.toJava))
+
+  def sourceRef[T](): javadsl.Sink[T, CompletionStage[SourceRef[T]]] =
+    scaladsl.Sink.sourceRef().mapMaterializedValue((_: Future[SourceRef[T]]).toJava).asJava
 }
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:285]: nice :)

**Comment by ktoso** [comment]
Great refactor, thanks :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 4]
```diff
@@ -602,4 +603,7 @@ object Source {
    */
   def unfoldResourceAsync[T, S](create: () ⇒ Future[S], read: (S) ⇒ Future[Option[T]], close: (S) ⇒ Future[Done]): Source[T, NotUsed] =
     Source.fromGraph(new UnfoldResourceSourceAsync(create, read, close))
+
+  def sinkRef[T](): Source[T, Future[SinkRef[T]]] =
+    Source.fromGraph(new RemoteStreamReceiverStage[T](OptionVal.None))
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:608]: I felt that's more confusing -- used: `SinkRef` / `SinkRefImpl` / `SinkRefStageImpl`

**Pull request merged**